libraries/chain/controller.cpp

```diff
@@ -2796,16 +2796,8 @@ struct controller_impl {
 
    // thread safe, expected to be called from thread other than the main thread
    block_state_legacy_ptr create_block_state_i( const block_id_type& id, const signed_block_ptr& b, const block_header_state_legacy& prev ) {
-      bool hs_active = false;
-      if (!b->header_extensions.empty()) {
-         std::optional<block_header_extension> instant_finality_ext = b->extract_header_extension(instant_finality_extension::extension_id());
-#warning change to use instant_finality_ext https://github.com/AntelopeIO/leap/issues/1508
-         if (instant_finality_ext) {
-            const auto& ext = std::get<instant_finality_extension>(*instant_finality_ext);
-            hs_active = !!ext.new_proposer_policy;
-         }
-      }
-
+      uint32_t hs_lib = hs_irreversible_block_num.load();
```

Comment:
We need a better way to do this. We need to make this determination via `prev` as timing of this call can happen right before `hs_irreversible_block_num` is set. We could just live with that as validation will fail, but it would fall back to trying again.

Reply:
How can `prev` be used to tell if current block `b` is constructed under instant finality. I think the fall back way is better -- if validation fail, we try the other way.